packages/ts/src/createTypeScriptFileFromProgram.ts

```diff
@@ -45,35 +52,53 @@ export function createTypeScriptFileFromProgram(
 					}),
 				}));
 		},
-		async runRule(rule, options) {
+		async runRule<MessageId extends string, FileContext extends object>(
+			runtime: RuleRuntime<
+				TSNodesByName,
+				MessageId,
+				TypeScriptServices,
+				FileContext
+			>,
+			messages: Record<string, ReportMessageData>,
+		): Promise<NormalizedReport[]> {
 			const reports: NormalizedReport[] = [];
 
-			const context = {
+			const services = {
 				program,
+				sourceFile,
+				typeChecker: program.getTypeChecker(),
+			};
+
+			const context = {
+				...services,
 				report: (report: RuleReport) => {
 					reports.push({
 						...report,
 						fix:
 							report.fix && !Array.isArray(report.fix)
 								? [report.fix]
 								: report.fix,
-						message: rule.messages[report.message],
+						message: messages[report.message],
 						range: normalizeRange(report.range, sourceFile),
 					});
 				},
-				sourceFile,
-				typeChecker: program.getTypeChecker(),
+				...(await runtime.fileSetup?.(services)),
 			};
 
-			const runtime = await rule.setup(context, options);
-
-			if (!runtime?.visitors) {
+			if (!runtime.visitors) {
 				return reports;
 			}
 
 			const { visitors } = runtime;
 			const visit = (node: ts.Node) => {
-				visitors[NodeSyntaxKinds[node.kind]]?.(node);
+				// TODO: There's got to be a better way to type visitors so all this casting isn't necessary.
+				const visitor = visitors[
+					NodeSyntaxKinds[node.kind] as keyof typeof visitors
+				] as
+					| RuleVisitor<typeof node, MessageId, TypeScriptServices>
+					| undefined;
+
+				visitor?.(node, context);
 
 				node.forEachChild(visit);
 			};
```

Comment:
Am I doing something wrong or will these all just need a cast? visitor is otherwise typed as `RuleVisitor<Node1Type, MessageId, TSServices> | ...<Node2Nype, ...> | ... | undefined`, which means that the node parameter gets typed as the intersection (`never`) because it has to be able to be passed to all of those, which it obviously can't be (it's a union).